src/sink.rs

```diff
@@ -0,0 +1,277 @@
+use alloc::vec::Vec;
+use core::mem::MaybeUninit;
+use core::ops::{Deref, DerefMut};
+
+/// Sink is used as target to de/compress data into a preallocated and possibly uninitialized memory space.
+/// Sink can be created from a `Vec` or a `Slice`. The new pos on the data after the operation
+/// can be retrieved via `sink.pos()`.
+///
+/// # Handling of Capacity
+/// Extend methods will panic if there's insufficient capacity left in the Sink.
+///
+/// # Safety invariants
+///   - `self.output[.. self.pos]` is always initialized.
+pub struct Sink<'a> {
+    /// The working slice, which may contain uninitialized bytes
+    output: &'a mut [MaybeUninit<u8>],
+    /// The Sink write position.
+    /// Also the number of bytes from the start of `output` guaranteed to be initialized.
+    pos: usize,
+}
+
+impl<'a> Sink<'a> {
+    /// Creates a `Sink` backed by the given byte slice.
+    #[inline]
+    pub fn new(output: &'a mut [u8], pos: usize) -> Self {
+        // SAFETY: Caller guarantees that all elements of `output` are initialized.
+        let _ = &mut output[..pos]; // bounds check pos
+        Sink {
+            output: slice_as_uninit_mut(output),
+            pos,
+        }
+    }
+}
+
+impl<'a> Sink<'a> {
+    /// Pushes a byte to the end of the Sink.
+    #[cfg(any(feature = "safe-encode", feature = "safe-decode"))]
+    #[inline]
+    pub fn push(&mut self, byte: u8) {
+        self.output[self.pos] = MaybeUninit::new(byte);
+        self.pos += 1;
+    }
+
+    /// Pushes `len` elements of `byte` to the end of the Sink.
+    /// # Panics
+    /// Panics if `copy_len` > `data.len()`.
+    #[cfg(any(feature = "safe-encode", feature = "safe-decode"))]
+    #[inline]
+    pub fn extend_with_fill(&mut self, byte: u8, len: usize) {
+        self.output[self.pos..self.pos + len].fill(MaybeUninit::new(byte));
+        self.pos += len;
+    }
+
+    /// Extends the Sink with `data` but only advances the sink by `copy_len`.
+    /// # Panics
+    /// Panics if `copy_len` > `data.len()`.
+    #[inline]
+    pub fn extend_from_slice_wild(&mut self, data: &[u8], copy_len: usize) {
+        // SAFETY: The assertion prevent Sink from exposing uninitialized data.
+        assert!(copy_len <= data.len());
+        self.output[self.pos..self.pos + data.len()].copy_from_slice(slice_as_uninit_ref(data));
```

Comment:
I think the straightforward way would be to take a ptr form slice and use that to copy

Reply:
Isn't this straightforward? We can go the ptr route but I needs extra lines to do the bounds checking.